Add quick-tune support for Attention

mirza-halilcevic opened the pull request:
## Motivation

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

Implement proper quick tuning support for Attention operations.

## Technical Details

<!-- Explain the changes along with any relevant GitHub links. -->

- Remove InitParams, and use Rock attributes directly instead
- Combine WmmaGemmParams and MfmaGemmParams into AccelGemmParams and implement deserialization
- Store serialized perfconfigs inside QuickTuningPerfconfigs.inc and update ParamLookupTable accordingly
- Rename AttnPerfConfigAttr to GemmGemmParams
- Rename other occurences of Attn to GemmGemm in order to match the Rock interface
- Introduce GridwiseGemmGemmParams.h/cpp with PopulateParamsGemmGemm implementation
- Refactor and update quickTuningGen.py

Resolves https://github.com/ROCm/rocMLIR-internal/issues/1887
Resolves https://github.com/ROCm/rocMLIR-internal/issues/1447

## Test Plan

<!-- Explain any relevant testing done to verify this PR. -->

## Test Result

<!-- Briefly summarize test outcomes. -->

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

mirza-halilcevic committed:
Add quick-tune support for Attention.

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/Tuning/GridwiseGemmGemmParams.h:
```diff
@@ -0,0 +1,145 @@
+namespace mlir {
+namespace rock {
+
+struct InitParamsAttn : Serializable<InitParamsAttn> {
```
I'd rather get rid of InitParams completely. There's an old ticket to do that https://github.com/ROCm/rocMLIR-internal/issues/1447

dhernandez0 replied:
I mean, there's no upside in using this rather than using `AttnPerfConfigAttr` directly, right? It's just duplicating the same thing twice.

mirza-halilcevic replied:
I don't see an upside; I was just trying to be consistent with the existing approach. It does seem to complicate things for no reason. I'll get rid of it.

dhernandez0 replied:
thanks!!

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/Tuning/QuickTuningPerfconfigs.inc:
```diff
@@ -1734,6 +1734,73 @@ static const InitParamsAccel initParametersForwardI8ConvGfx1150[nInitParametersF
+
+// BEGIN_ATTENTION_XDL_f32_gfx900_DEFS
+const InitParamsAttn PopulateParamsAttnXDL::initParametersAttentionGfx900[PopulateParamsAttnXDL::nInitParametersAttentionGfx900] = {
+    {32,128,64,32,32,32,16,4,1,1,2,true},
```
is this the existing list?

mirza-halilcevic replied:
Yes, for now. This will be updated.

mirza-halilcevic committed:
Remove usage of InitParamsAttn and simplify approach.

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/Tuning/QuickTuningPerfconfigs.inc:
nit: own file for attention/g+g?

dhernandez0 commented on mlir/lib/Dialect/Rock/Tuning/GridwiseGemmGemmParams.cpp:
```diff
@@ -0,0 +1,133 @@
+std::vector<AttnPerfConfigAttr>
+PopulateParamsAttn::getQuickTuningRange(OpBuilder &b,
+                                        RockGemmGemmWrapperInterface op) {
+  if (!bitEnumContainsAny(rock::getFeatures(op),
```
rock::isAccel()

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/Tuning/GridwiseGemmGemmParams.h:
```diff
@@ -0,0 +1,68 @@
+
+class PopulateParamsAttn {
+public:
+  struct PerfConfig {
```
we could do this: `AttnPerfConfigAttr::get(perfConfigString, isWmma);` instead of passing numbers

dhernandez0 commented on mlir/lib/Dialect/Rock/Tuning/GridwiseGemmGemmParams.cpp:
```diff
@@ -0,0 +1,133 @@
+      params.getForceUnroll());
+}
+
+template <typename GemmParamsAttrType>
```
can we use `RockAccelTuningParamAttrInterface` instead? It is the parent class of both mfma and wmma

dhernandez0 commented on mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp:
```diff
@@ -240,31 +241,24 @@ getAccelRangeAttn(RockGemmGemmWrapperInterface gemmGemmOp,
-      rock::lookupArchInfo(rock::getArchValue(gemmGemmOp)).waveSize;
-  if (!bitEnumContainsAny(features, GemmFeatures::mfma) &&
-      !bitEnumContainsAny(features, GemmFeatures::wmma)) {
+  if (!bitEnumContainsAny(rock::getFeatures(gemmGemmOp),
```
rock::isAccel

dhernandez0 commented on mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp:
```diff
@@ -316,21 +311,19 @@ static void createAttnTuningRangeGreedyPhase1(
   GemmFeatures features = rock::getFeatures(gemmGemmOp);
-  bool isWMMA = bitEnumContainsAny(features, GemmFeatures::wmma);
-  if (!bitEnumContainsAny(features, GemmFeatures::mfma) && !isWMMA) {
+  if (!bitEnumContainsAny(features, GemmFeatures::mfma | GemmFeatures::wmma)) {
```
rock::isAccel

dorde-antic committed:
Merge branch 'develop' into attn-quick-tune

mirza-halilcevic committed:
Remove InitParamsAttn and store serialized perf configs in inc file.

mirza-halilcevic committed:
Merge remote-tracking branch 'origin/develop' into attn-quick-tune

mirza-halilcevic committed:
Merge remote-tracking branch 'origin/attn-quick-tune' into attn-quick-tune

mirza-halilcevic committed:
Merge remote-tracking branch 'origin/develop' into attn-quick-tune

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/Tuning/QuickTuningPerfconfigs.inc:
```diff
@@ -1802,3 +1861,14 @@ static const InitParamsAccel initParametersForwardI8ConvGfx1150[nInitParametersF
+
+#ifdef Attn_LOOKUP_TABLE_GEN
+
+{"gfx900_attention_f32", {PopulateParamsAttn::initParametersAttentionGfx900, PopulateParamsAttn::nInitParametersAttentionGfx900}},
```
I guess temporarily, gemm+gemm kernels will use attention quick tuning list but eventually, we'll have a list for gemm_gemm? (and potentially conv+gemm). We already have a tier1-gemmgemm list

mirza-halilcevic replied:
Yes, the ticket is in progress https://github.com/ROCm/rocMLIR-internal/issues/2019

mirza-halilcevic committed:
Store quick-tune perf configs in serialized format.

mirza-halilcevic committed:
Refactor tuning parameters:
- Retire InitParamsNonAccel and InitParamsAccel
- Combine MfmaGemmParamsAttr and WmmaGemmParamsAttr into
  AccelGemmParamsAttr

mirza-halilcevic committed:
Merge remote-tracking branch 'origin/develop' into attn-quick-tune

mirza-halilcevic committed:
Fix tests.

dhernandez0 commented on mlir/include/mlir/Dialect/Rock/Tuning/GridwiseGemmGemmParams.h:
```diff
@@ -0,0 +1,62 @@
+
+class PopulateParamsAttn {
+public:
+  static std::vector<AttnParamsAttr>
```
maybe this is a good opportunity to rename it `GemmGemmParamsAttr`? to be consistent with the rest `RockGemmGemmWrapperInterface`, `getGemmGemmTuningParams`, etc...

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/RockDialect.cpp:
```diff
@@ -3351,90 +3351,210 @@ void AttentionOp::getEffects(
+  int64_t mPerWave = params[idx++];
+
+  int64_t nPerWave, mnPerXdl;
+  if (version > 3) {
```
nit: refactor all these if/else if/else into a function (used here and for attention)

dhernandez0 commented on mlir/lib/Dialect/Rock/Tuning/GridwiseGemmGemmParams.cpp:
```diff
@@ -0,0 +1,114 @@
+using namespace mlir;
+using namespace mlir::rock;
+
+#define Attn_DEFINITIONS_GEN
```
Attn -> GemmGemm?

dhernandez0 commented on mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp:
```diff
@@ -691,116 +684,62 @@ static void createGemmTuningRangeBF(TuningParamSet *newSpace,
-            cast<RockTuningParamAttrInterface>(attnParams));
-      }
+  for (AttnParamsAttr params :
+       PopulateParamsAttn::getQuickTuningRange(b, gemmGemmOp)) {
```
PopulateParamsAttn -> PopulateParamsGemmGemm?

mirza-halilcevic committed:
Fix formatting.

mirza-halilcevic committed:
Merge branch 'develop' into attn-quick-tune

mirza-halilcevic committed:
Merge remote-tracking branch 'origin/attn-quick-tune' into attn-quick-tune

mirza-halilcevic committed:
Merge remote-tracking branch 'origin/develop' into attn-quick-tune

dhernandez0 committed:
Merge branch 'develop' into attn-quick-tune

mirza-halilcevic committed:
Merge remote-tracking branch 'origin/develop' into attn-quick-tune

mirza-halilcevic committed:
Rename Attn to GemmGemm for generic constructs.

mirza-halilcevic committed:
Improve naming.

mirza-halilcevic committed:
Update and refactor quickTuningGen.py.

mirza-halilcevic committed:
Merge branch 'develop' into attn-quick-tune

mirza-halilcevic committed:
Merge branch 'develop' into attn-quick-tune

mirza-halilcevic committed:
Merge branch 'develop' into attn-quick-tune

mirza-halilcevic committed:
Merge branch 'develop' into attn-quick-tune

dhernandez0 commented on mlir/lib/Dialect/Rock/IR/RockDialect.cpp:
```diff
@@ -3351,90 +3351,215 @@ void AttentionOp::getEffects(
+    mPerWave = mPerBlock / mWaves;
+    nPerWave = std::max(nPerBlock / nWaves, mnPerXdl);
+  }
+  return {nPerWave, mnPerXdl};
```
mPerWave might change as well in the else branch, right?

dhernandez0 replied:
oh, I see it's passed as reference? `int64_t &mPerWave`. I'd prefer to return it if possible or at least add a comment to highlight it.

dhernandez0 commented on mlir/lib/Dialect/Rock/utility/AccelEmitter.cpp:
```diff
@@ -1296,7 +1296,7 @@ AccelEmitter::select(GemmFeatures features, Type dataTypeA, Type dataTypeB,
                                          tuningParams);
   } else if (isWmma) {
-    WmmaGemmParamsAttr wmmaParams = cast<WmmaGemmParamsAttr>(tuningParams);
+    AccelGemmParamsAttr wmmaParams = cast<AccelGemmParamsAttr>(tuningParams);
```
do we need to cast this? I understand tuningParams is already `AccelGemmParamsAttr`?

mirza-halilcevic replied:
You're right, we don't. This was the result of a find and replace so I missed it. Thanks.

mirza-halilcevic replied:
Actually, tuningParams is `RockAccelTuningParamAttrInterface`, so we are downcasting.

mirza-halilcevic committed:
Merge remote-tracking branch 'origin/develop' into attn-quick-tune

mirza-halilcevic committed:
Explicitly return mPerWave when parsing instead of taking it by
reference.

mirza-halilcevic committed:
Merge remote-tracking branch 'origin/attn-quick-tune' into attn-quick-tune
